prereise/gather/demanddata/transportation_electrification/charging.py

```diff
@@ -0,0 +1,121 @@
+def get_charging_power(power, trip_strategy, location_strategy, kwh, trip_data):
+    '''Determines the charging power. Returns either param power or 0.
+
+    :param int power: charger power, EVSE kW.
+    :param int trip_strategy: a flag that determines which trips to consider, 1-anytrip number, 2-last trip.
+    :param int location_strategy: where the vehicle can charge-1, 2, 3, 4, or 5; 1-home only, 2-home and work related, 3-anywhere if possibile, 4-home and school only, 5-home and work and school.
+    :param float kwh: kwhmi * veh_range, amount of energy needed to charge vehicle.
+    :param pandas.Series trip_data: Row of DataFrame that has data for the trip we are calculating for.
+    :return: (*float*) -- charging power.
+    '''
+    charging_power = power
+
+    dwelling = int(trip_data['Dwell time (hour decimal)'] > 0.2)
+    location = get_location(location_strategy, trip_data['why to'])
+
+    trip = consider_trip_number(trip_strategy, trip_data['total vehicle trips'], trip_data['trip number'])
+    
+    battery = get_battery_SOC(trip_data['trip end battery charge'], kwh) 
+    if dwelling * location * trip * battery != 1 and not (dwelling * location == 1 or trip_data['total vehicle trips'] == trip_data['trip number']):
+        charging_power = 0
+    
+    return charging_power
+
+
+def get_charging_time(charging_power, kwh, battery_SOC, charging_efficiency):
+    '''Calculates time needed to charge.
+
+    :param float charging_power: charging power.
+    :param float kwh: kwhmi * veh_range, amount of energy needed to charge vehicle.
+    :param float battery_SOC: battery charge at end of trip.
+    :param float charging_efficiency: grid to battery efficiency.
+    :return: (*float*) -- charging time in decimal format.
+    '''
+    if charging_power == 0:
+        charging_time = 0
+    else:
+        charging_time = (kwh - battery_SOC)/charging_power/charging_efficiency

```

Comment:
I think this would be clearer as `charging_time = (kwh - battery_SOC)/(charging_power * charging_efficiency)`.